### PR TITLE
feat(cass): Consolidate to one partkey table instead of table per shard

### PR DESCRIPTION
**Pull request by vishramachandran**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Today, we have one table per shard for storing part keys.
This (work-in-progress) PR explores storing all part keys in one table. 
This is to reduce Cassandra compaction cost incurred due to numerous tables.

All features hidden behind feature flag.

These are the read/write path considerations in the design of Cassandra schema
* should be able to load contents of shard efficiently for index bootstrap
* should be able to read/write single part key efficiently for regular ingestion operations
* Should be able to do cardinality busting
* Should be able to do index migration for downsampling
* Should be able to repair across DCs

TODOs for later:
* Performance Testing
* Data migration plan

### CODE REVIEW

**Review by alextheimer — APPROVED** [review]
There's some copy-paste going on here, but I don't think there's any better way to approach this frequent if-else-with-similar-blocks logic.
- We could write generalized, complex-parameter functions to prevent copy-paste, but that would complicate readability.
- We could write classes to handle the if/else statements via dynamic dispatch, but that's too much code/complexity for a flag we intend to keep temporary.

This looks good 👍

File: cassandra/src/main/scala/filodb.cassandra/columnstore/PartitionKeysV2Table.scala
Comment (amolnayak311) [inline cassandra/src/main/scala/filodb.cassandra/columnstore/PartitionKeysV2Table.scala:18]: Will have to introduce read consistency similar to what got introduced in PR https://github.com/filodb/FiloDB/pull/1573

File: core/src/main/scala/filodb.core/store/ChunkSink.scala
Comment (amolnayak311) [inline core/src/main/scala/filodb.core/store/ChunkSink.scala:23]: Nit pick, Perhaps an assert on numBuckets to be non zero?

File: core/src/test/scala/filodb.core/memstore/ratelimit/RocksDbCardinalityStoreMemoryCapSpec.scala
Comment (amolnayak311) [inline core/src/test/scala/filodb.core/memstore/ratelimit/RocksDbCardinalityStoreMemoryCapSpec.scala:51]: Not a big fan of putting assertions on runtimes in Spec these may lead to flaky tests based on the different platform.

**Review by amolnayak311 — COMMENTED** [review]
Thanks a lot for the changes and apologies for taking long to review. It seems you need to rebase and some changes around read consistency will be needed. I am happy approve once the changes are pushed as approval will be needed anyways once a new commit is made..

**Review by amolnayak311 — COMMENTED** [review]
Thanks a lot for the changes and apologies for taking long to review. It seems you need to rebase and some changes around read consistency will be needed. I am happy approve once the changes are pushed as approval will be needed anyways once a new commit is made..

**Comment by whizkido** [comment]
Finally !!!!!

File: cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala
Comment (vishramachandran) [reply cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala]: Note that `getOrCreatePartitionKeysV2Table` does not actually create the table in cassandra. It creates the `PartitionKeysV2Table` object. The `initialize` call will create the tables, and is lazy.

I did not add the guards for v1 tables because we did not narrow down on migration strategy. We may choose to migrate with co-existing tables. That is an option on the table.

Nevertheless, it is likely we may not choose the co-existing tables strategy for migration, so I will disable the table object initialization anyway.

File: cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala
Comment (vishramachandran) [reply cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala:85]: Again, table creation happens only on `intialize()` call, not upon creating the handle object.

File: cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala
Comment (vishramachandran) [reply cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala:533]: See other comments, but done anyway

File: cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala
Comment (vishramachandran) [reply cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala:350]: Spread changes should not be clubbed with migration or chunk repair. Anyway, this is out of scope for this PR.

File: cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala
Comment (vishramachandran) [reply cassandra/src/main/scala/filodb.cassandra/columnstore/CassandraColumnStore.scala:690]: Table creation will not happen, but added anyway to catch any issues.

File: cassandra/src/main/scala/filodb.cassandra/columnstore/PartitionKeysV2Table.scala
Comment (vishramachandran) [reply cassandra/src/main/scala/filodb.cassandra/columnstore/PartitionKeysV2Table.scala:18]: Yes - This is done now after rebase.

File: core/src/main/scala/filodb.core/store/ChunkSink.scala
Comment (vishramachandran) [reply core/src/main/scala/filodb.core/store/ChunkSink.scala:23]: This is unlikely to happen since we have default configs. Even if it happens, an IllegalArg is not any better than DivideByZero exception. I'll leave it as is.

**Review by amolnayak311 — APPROVED** [review]
Thanks @vishramachandran I missed the ``getOrCreate`` object vs create table.  I think it's good to go. Thanks for the quick turnaround

**Pull request merged**
